File: apps/web/lib/api/items.ts
import type { paths } from "@ims/api-types"
import { apiFetch } from "./client"

type ItemRead = paths["/api/items/{item_id}"]["get"]["responses"]["200"]["content"]["application/json"]
type ItemListResponse = paths["/api/items"]["get"]["responses"]["200"]["content"]["application/json"]
type ItemCreate = paths["/api/items"]["post"]["requestBody"]["content"]["application/json"]
type ItemUpdate = paths["/api/items/{item_id}"]["patch"]["requestBody"]["content"]["application/json"]

export type { ItemRead, ItemListResponse, ItemCreate, ItemUpdate }

export interface ItemFilters {
  q?: string
  categoryId?: number
  locationId?: number
  tagIds?: number[]
  page?: number
  perPage?: number
}

function buildQuery(filters: ItemFilters): string {
  const params = new URLSearchParams()
  if (filters.q) params.set("q", filters.q)
  if (filters.categoryId != null) params.set("category_id", String(filters.categoryId))
  if (filters.locationId != null) params.set("location_id", String(filters.locationId))
  if (filters.tagIds && filters.tagIds.length > 0) {
    for (const id of filters.tagIds) params.append("tag_ids", String(id))
  }
  if (filters.page != null) params.set("page", String(filters.page))
  if (filters.perPage != null) params.set("per_page", String(filters.perPage))
  const s = params.toString()
  return s ? `?${s}` : ""
}

export async function listItems(filters: ItemFilters, accessToken: string | null): Promise<ItemListResponse> {
  const res = await apiFetch(`/items${buildQuery(filters)}`, { accessToken })
  return (await res.json()) as ItemListResponse
}

export async function getItem(id: string, accessToken: string | null): Promise<ItemRead> {
  const res = await apiFetch(`/items/${id}`, { accessToken })
  return (await res.json()) as ItemRead
}

export async function createItem(body: ItemCreate, accessToken: string | null): Promise<ItemRead> {
  const res = await apiFetch("/items", {
    method: "POST", body: JSON.stringify(body), accessToken,
  })
  return (await res.json()) as ItemRead
}

export async function updateItem(id: string, body: ItemUpdate, accessToken: string | null): Promise<ItemRead> {
  const res = await apiFetch(`/items/${id}`, {
    method: "PATCH", body: JSON.stringify(body), accessToken,
  })
  return (await res.json()) as ItemRead
}

export async function deleteItem(id: string, accessToken: string | null): Promise<void> {
  await apiFetch(`/items/${id}`, { method: "DELETE", accessToken })
}

File: apps/api/app/routes/items.py
from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, Depends, Query, Response, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth.dependencies import get_current_user
from app.db.session import get_db
from app.models.user import User
from app.schemas.item import ItemCreate, ItemListResponse, ItemRead, ItemUpdate
from app.services import items_service

router = APIRouter(prefix="/api/items", tags=["items"])


@router.get("", response_model=ItemListResponse)
async def list_items(
    q: str | None = None,
    category_id: int | None = None,
    location_id: int | None = None,
    tag_ids: list[int] | None = Query(default=None),
    favorite: bool | None = None,
    page: int = Query(default=1, ge=1),
    per_page: int = Query(default=20, ge=1, le=100),
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> ItemListResponse:
    return await items_service.list_items(
        session,
        user.id,
        q=q,
        category_id=category_id,
        location_id=location_id,
        tag_ids=tag_ids,
        favorite=favorite,
        page=page,
        per_page=per_page,
    )


@router.post("/{item_id}/favorite", response_model=ItemRead)
async def toggle_favorite(
    item_id: UUID,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> ItemRead:
    return await items_service.toggle_favorite(session, user.id, item_id)


@router.post("", response_model=ItemRead, status_code=status.HTTP_201_CREATED)
async def create_item(
    body: ItemCreate,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> ItemRead:
    return await items_service.create_item(session, user.id, body)


@router.get("/{item_id}", response_model=ItemRead)
async def get_item(
    item_id: UUID,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> ItemRead:
    return await items_service.get_item(session, user.id, item_id)


@router.patch("/{item_id}", response_model=ItemRead)
async def update_item(
    item_id: UUID,
    body: ItemUpdate,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> ItemRead:
    return await items_service.update_item(session, user.id, item_id, body)


@router.delete("/{item_id}", status_code=status.HTTP_204_NO_CONTENT, response_class=Response)
async def delete_item(
    item_id: UUID,
    user: User = Depends(get_current_user),
    session: AsyncSession = Depends(get_db),
) -> Response:
    await items_service.delete_item(session, user.id, item_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)
